File: src/cxx-dev/LiENa/lienaDistributedSystem/lienaDistributedSystem.cpp
#include "lienaDistributedSystem.h"


/**
 *
 * @brief lienaDistributedSystem::lienaDistributedSystem
 * @param globalParameter
 *
 */
lienaDistributedSystem::lienaDistributedSystem(lienaGlobal *globalParameter){
    this->globalParameter = globalParameter;
    this->init();
}

//! -----------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDistributedSystem::~lienaDistributedSystem
//!
lienaDistributedSystem::~lienaDistributedSystem(){
    delete this->globalParameter;
    delete &this->distributedModules;
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDistributedSystem::checkModuleByAddr
//! \param addr
//! \return
//!
bool lienaDistributedSystem::checkModuleByAddr(QString addr){
    bool ret = false;
    for(int i = 0; i < distributedModules.size(); i++){
        if(distributedModules.at(i)->getAddress() == addr){
            ret = true;
        }
    }
    return ret;
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDistributedSystem::configureMessageQueuePairBy
//! \param deviceID
//! \param outputMessageQueue
//! \param inputMessageQueue
//!
void lienaDistributedSystem::configureMessageQueuePairBy(uint32_t deviceID, lienaMessageQueue *outputMessageQueue,lienaMessageQueue *inputMessageQueue){
    qDebug()<<"lienaDistributedSystem::configureMessageQueuePairBy";
    for(int i = 0; i < distributedModules.size(); i++){
        if(distributedModules.at(i)->getDeviceId() == deviceID){
            distributedModules.at(i)->configureMessageQueuePair(outputMessageQueue,inputMessageQueue);
        }
    }
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDistributedSystem::generateReceptionChannelByAddr
//! \param socketReception
//! \param addr
//!
void lienaDistributedSystem::generateReceptionChannelByAddr(SOCKET socketReception, QString addr){
    for(int i  = 0; i < distributedModules.size(); i++){
        if(distributedModules.at(i)->getAddress() == addr){
            if(distributedModules.at(i)->isInDiagnosis()){
                distributedModules.at(i)->repairReceptionChannel(socketReception);
            }
            else{
                distributedModules.at(i)->generateReceptionChannel(socketReception);
            }
        }
    }
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//!
//! \brief lienaDistributedSystem::setRoboticSystemModule
//! \param deviceIDs
//!
void lienaDistributedSystem::setRoboticSystemModule(QVector<uint32_t> deviceIDs){
    this->globalDeviceIDs = deviceIDs;
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDistributedSystem::creatIncomingModule
//! \return
//!
void lienaDistributedSystem::createDistributedModuleForReceptionChannel(SOCKET socketReception){
    lienaDistributedModule *module = new lienaDistributedModule(this->moduleIndex, this->globalParameter);
    this->connect(module, SIGNAL(lostConnection(uint32_t)), this, SLOT(launchRecoveryProcedure(uint32_t)));
    this->connect(module, SIGNAL(generateNewMessageSequence(uint32_t)), this, SLOT(notifyGenerateNewMessageSequence(uint32_t)));

    module->generateReceptionChannel(socketReception);
    this->distributedModules.append(module);
    this->moduleIndex ++;
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDistributedSystem::launchRecoveryProcedure
//! \param deviceId
//!
void lienaDistributedSystem::launchRecoveryProcedure(uint32_t deviceId){
    qDebug()<<"lienaDistributedSystem::launchRecoveryProcedure"<<deviceId;
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDistributedSystem::notifyGenerateNewMessageSequence
//! \param deviceId
//!
void lienaDistributedSystem::notifyGenerateNewMessageSequence(uint32_t deviceId){
    emit generateNewMessageSequence(deviceId);
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief createDistributedModuleWithTransmissionChannel
//! \param addr
//! \param port
//!
bool lienaDistributedSystem::createDistributedModuleWithTransmissionChannel(uint32_t targetDeviceId,bool motivated, QString addr, unsigned short port){

    lienaDistributedModule *module = new lienaDistributedModule(this->moduleIndex, this->globalParameter);

    this->connect(module, SIGNAL(lostConnection(uint32_t)), this, SLOT(launchRecoveryProcedure(uint32_t)));

    module->setTargetDeviceId(targetDeviceId);

    int ret = module->generateTransmissionChannel(motivated, addr, port);

    if(ret == 1){
        qDebug()<<"successful create a distributed module"<<targetDeviceId<<"with a transmission channel";
        this->distributedModules.append(module);
        this->moduleIndex ++;
        return true;
    }
    else{
        qDebug()<<"failed create a distributed module"<<targetDeviceId<<"with a transmission channel";
        qDebug()<<ret;
        return false;
    }
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDistributedSystem::closeDistributedChannel
//! \param deviceId
//!
void lienaDistributedSystem::closeDistributedChannel(uint32_t deviceId){

    for(int i = 0;i<this->distributedModules.count(); i++){
        lienaDistributedModule* mdles = this->distributedModules.at(i);
        if(mdles->getDeviceId() == deviceId){
            qDebug()<<"lienaDistributedSystem |closeDistributedChannel deviceId:"<<deviceId;
            mdles->closeDistributedChannel();
        }
    }
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDistributedSystem::init
//!
void lienaDistributedSystem::init(){
    this->clear();
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDistributedSystem::clear
//!
void lienaDistributedSystem::clear(){
    this->moduleIndex = 0;
    this->distributedModules.clear();
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDistributedSystem::append
//! \param module
//!
void lienaDistributedSystem::append(lienaDistributedModule*module){
    this->distributedModules.append(module);
}

//! ------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDistributedSystem::getModuleCount
//! \return
//!
int lienaDistributedSystem::getModuleCount(){
    return this->distributedModules.size();
}

//! -----------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDistributedSystem::getModuleIndexByID
//! \return
//!
int lienaDistributedSystem::getModuleIndexByID(int deviceId){
    int ret = -1;
    int len = getModuleCount();
    for(int i =0; i< len; i++){
        if(this->distributedModules.at(i)->getDeviceId() == deviceId){
            ret = this->distributedModules.at(i)->getIndex();
        }
    }
    return ret;
}

//! -------------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDistributedSystem::setDeiveIDByIndex
//! \param index
//! \param deviceId
//!
void lienaDistributedSystem::setDeiveIDByIndex(int index, int deviceId){
    this->distributedModules.at(index)->setTargetDeviceId(deviceId);
}

File: src/py-dev/LiENa/LiENaDistributedSystem/lienaDistributedSystem.py
from LiENaDistributedSystem.lienaDistributedModule import LienaDistributedModule
from LiENaBasic.lienaDefinition import *

from PyQt5.QtCore import QObject, pyqtSlot, pyqtSignal


class LienaDistributedSystem(QObject):

    needToRestartServer = pyqtSignal()
    generateNewMessageSequence = pyqtSignal(int)

    def __init__(self, global_parameter):
        super(LienaDistributedSystem, self).__init__()
        self.globalParameter = global_parameter
        self.moduleIndex = 0
        self.devicesIDs = list()
        self.distributedModules = list()

    def launch_recovery_procedure(self, device_id):
        print("LienaDistributedSystem | launch_recovery_procedure", device_id)

    def check_module_by_addr(self, addr):
        ret = False
        for module in self.distributedModules:
            if module.get_address() == addr:
                ret = True
        return ret

    def close_session_by_device_id(self, device_id):
        for module in self.distributedModules:
            if module.get_device_id() == device_id:
                module.close_session_request(device_id)

    def set_robotic_system_module(self, devices_ids):
        self.devicesIDs = devices_ids

    def generate_reception_channel_by_addr(self, addr, socket_for_reception):
        if DEBUG:
            print("LienaDistributedSystem | generate_reception_channel_by_addr")
        for module in self.distributedModules:
            if module.get_address() == addr:
                if module.is_in_diagnosis():
                    module.repair_reception_channel(socket_for_reception)
                else:
                    module.generate_reception_channel(socket_for_reception)

    def create_distributed_module_with_reception_channel(self, socket_for_reception):
        if DEBUG:
            print("LienaDistributedSystem | create_distributed_module_with_reception_channel")
        module = LienaDistributedModule(self.moduleIndex, self.globalParameter)
        module.lostConnexion[int].connect(self.launch_recovery_procedure)

        module.generateNewMessageSequence[int].connect(self.notify_generate_new_msg_seg)

        module.generate_reception_channel(socket_for_reception)
        self.distributedModules.append(module)
        self.moduleIndex += 1

    def configure_msg_sequence_pair_by(self, device_id, outputMegQue, inputMsgQue):
        print ("configure_msg_sequence_pair_by", device_id)
        for i in range(len(self.distributedModules)):
            if self.distributedModules[i].get_device_id() == device_id:
                self.distributedModules[i].configure_msg_Queue_Pair(outputMegQue, inputMsgQue)

    def notify_generate_new_msg_seg(self, device_id):
        self.generateNewMessageSequence.emit(device_id)

    def restart_server(self):
        self.needToRestartServer.emit()

    def create_distributed_module_with_transmission_channel(self, target_device_id, motivate, addr, port):
        module = LienaDistributedModule(self.moduleIndex, self.globalParameter)

        module.lostConnexion[int].connect(self.launch_recovery_procedure)
        module.generateNewMessageSequence[int].connect(self.notify_generate_new_msg_seg)
        module.restartServer.connect(self.restart_server)

        module.set_device_id(target_device_id)
        ret = module.generate_transmission_channel(motivate, addr, port)

        if ret == 0:
            if DEBUG:
                print("LienaDistributedSystem | successful create a distributed module with transmission channel")
            self.distributedModules.append(module)
            self.moduleIndex += 1
            return True
        else:
            if DEBUG:
                print("LienaDistributedSystem | failed create a distributed module with transmission channel")
            return False

    def clear(self):
        self.moduleIndex = 0
        self.distributedModules.clear()

    def append(self, moudle):
        self.distributedModules.append(moudle)

    def get_module_count(self):
        return self.distributedModules.size()

    def get_module_index_by_id(self, device_id):
        ret = 1
        len = self.get_module_count()
        for i in len:
            if self.distributedModules[i].get_device_id == device_id:
                ret = self.distributedModules[i].get_index()
        return ret

    def set_device_id_by_index(self, index, device_id):
        self.distributedModules[index].set_device_id(device_id)
